---
title: "Exam 4; Exam 3 Revisions"
author: "Alyson DeNittis"
date: "April 27, 2018"
---


### Part 1 (50 points): You have two fastq files in the Exam3 directory.

Load libraries

```{r eval=FALSE, include=TRUE}
library(ShortRead)
library(tidyr)
library(stringr)
library(msa)
library(seqinr)
library(dada2)
library(vegan)
library(ggplot2)
```

1. Convert these fastq reads to fasta reads. These new fasta reads should be named "Sample1.fasta" and "Sample2.fasta"

BASH CODE:
```{r eval=FALSE, include=TRUE}
alias fqtofa='sed -n '\''1~4s/^@/>/p;2~4p'\''''' 
for fn in *.fastq; do fqtofa $fn > $fn.fasta;done
```

2. How many DNA reads were detected in each sample?

Sample1.fastq.fasta:6577
Sample2.fastq.fasta:7731

BASH CODE:
```{r eval=FALSE, include=TRUE}
grep -c "^>" *.fasta
```

3. Trim the original fastq files so that only the first 100 bases are included from each DNA read within.
4. Save these trimmed fastq files as "Sample1.fastq.trim" and "Sample2.fastq.trim"

```{r, eval = FALSE}
fq.files = dir(path = getwd(), full.names = TRUE, pattern = ".fastq")
trim.files = dir(path = getwd(), full.names = FALSE, pattern = ".fastq")
dir.create(file.path(getwd(), "trimmed"))

for(i in filt.files){
  fastqFilter(fn = i,fout = paste0(getwd(),"/trimmed/",i,".trim"),
              truncLen = 100)
}
```

### Part 2 (50 points): There are two csv files in the Exam3 directory.
+ The file "exam3_otu_table.csv" is a "species" abundance table like we've seen before. It has the number of times a given DNA sequence was found in a given sample.
+ The file "exam3_metadata.csv" has information about the samples in the species abundance table.

I'm curious about whether the bacterial communities found in Soil are statistically different from those found in the Rhizosphere, **and** whether "Location" makes a difference (*i.e.*, is interactive).
I'm **not** interested in whether the host plant Species or Season are important

1. Import these files into R objects

```{r}
meta = read.csv("C:/Users/Aly/Desktop/GIT/Data_Course/Exams/Exam_3/exam3_metadata.csv")
otu = read.csv("C:/Users/Aly/Desktop/GIT/Data_Course/Exams/Exam_3/exam3_otu_table.csv")
```
```{r}
print(head(meta,5))
print(head(otu,3))
```
2. Subset both objects to include only samples where SampleType is either "Soil" or "rhizosphere"

```{r}
meta = read.csv("C:/Users/Aly/Desktop/GIT/Data_Course/Exams/Exam_3/exam3_metadata.csv")
otu = read.csv("C:/Users/Aly/Desktop/GIT/Data_Course/Exams/Exam_3/exam3_otu_table.csv")
otunames = sub("_","",row.names(otu))
rownames(otu) = otunames
metanames = sub("_","",row.names(meta))
metanames = rownames(meta) 

sampt = subset(meta, meta$SampleType == "Soil" | meta$SampleType == "rhizosphere")
subotu = as.data.frame(otu[,names(otu) %in% sampt$SampleID])

t_otu = as.data.frame(t(subotu))


```

3. Perform a PermANOVA to determine whether there are significant differences in bacterial community composition between these two sample types.

 

```{r include=FALSE}
library(vegan)
```
```{r}
adonis(t_otu ~ sampt$SampleType*sampt$Location)

```

Based on these results, there seems to be a significant difference between location and community composition, but not on the sample type or the interaction between sample type and location.

4. Determine which bacterial *genus* is the **most abundant** for each of those sample types as a whole
Note: Genus is denoted by "g__" in the "Consensus.lineage" column of the species abundance table

```{r}
soilsub = subset(sampt, SampleType == "Soil")
rhizosub = subset(sampt, SampleType == "rhizosphere")
sotu = subotu[,which(names(subotu) %in% soilsub$SampleID)]
rotu = subotu[,which(names(subotu) %in% rhizosub$SampleID)]
smax = max(rowSums(sotu))
rmax = max(rowSums(rotu))
smax2 = which(rowSums(sotu) == smax)
rmax2 = which(rowSums(rotu) == rmax)
masoil = otu$Consensus.lineage [smax2]
marhiz = otu$Consensus.lineage[rmax2]
print(masoil)
print(marhiz)
```

The genus most abundant for the soil sample is (possibly) _Bacillales_, and the genus most abundant in rhizosphere sample is (possibly) _Bacillus_.I spent a lot of time with different split and grep functions to separate the vector, but I couldn't get it right. 
